### PR TITLE
Remove support for DDP2 strategy

### LINKED ISSUE #12584: RFC: Deprecate DataParallel2 Strategy (DDP2)
## Proposed refactor

Deprecate the DDP2 strategy.

### Motivation

We have few to no users. Evidence:

- The strategy hasn't been working for at least two PL versions
- It is broken for all PyTorch versions >= 1.9 and it is unclear how to make it work for PyTorch >= 1.9. [Explanation](https://github.com/PyTorchLightning/pytorch-lightning/issues/9667#issuecomment-986161540)
- We get no questions, bug reports, and no feature requests related to DDP2
- There are no functional tests for DDP2

Hence, it is not worth the maintenance effort to keep DDP2, especially since it is broken.

### Pitch

Deprecate and remove DDP2 in PL 1.9.

### Additional context

Closes issues #9667

Code repository search shows that the only usages are from within PL and forks of PL:
https://grep.app/search?q=DDP2Plugin
https://github.com/search?p=1&q=DDP2Plugin&type=Code

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @tchaton

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #12584 

Raises an error for users who subclass or use the DDP2Strategy/DDP2Plugin. Raises an error for everyone trying to use it with `strategy="ddp2"`. 

### Does your PR introduce any breaking changes? If yes, please list them.

Yes. If you try this, you get an error:

```py
trainer = Trainer(strategy="ddp2")
```
Since DDP2 is broken and unsupported on torch >= 1.9, you will need to downgrade Lightning to < 1.5 with torch < 1.9. 
It is possible to instantiate the DDP2Strategy and pass it into the strategy argument, but it will still print a deprecation warning. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @tchaton @rohitgr7 @borda @justusschock @awaelchli

### CODE REVIEW

File: docs/source/accelerators/gpu.rst [hunk 1]
```diff
@@ -765,28 +771,14 @@ In DDP, DDP_SPAWN, Deepspeed, DDP_SHARDED, or Horovod your effective batch size
     Trainer(accelerator="gpu", devices=8, num_nodes=10, strategy="ddp_sharded")
     Trainer(accelerator="gpu", devices=8, num_nodes=10, strategy="horovod")
 
-In DDP2 or DP, your effective batch size will be 7 * num_nodes.
-The reason is that the full batch is visible to all GPUs on the node when using DDP2.
```
Comment (awaelchli) [inline docs/source/accelerators/gpu.rst:769]: todo: rewrite this for DP only.

File: pytorch_lightning/strategies/ddp2.py [hunk 2]
```diff
@@ -16,15 +16,31 @@
...
+        recommend the :class:`~pytorch_lightning.strategies.ddp.DDPStrategy` or the
+        :class:`~pytorch_lightning.strategies.dp.DataParallelStrategy` as a replacement. If you need DDP2, you will
+        need ``torch < 1.9`` and ``pytorch-lightning < 1.5``.
+    """
 
     strategy_name = "ddp2"
 
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        rank_zero_deprecation(
```
Comment (carmocca) [inline pytorch_lightning/strategies/ddp2.py:38]: I would show the deprecation before calling super().__init__

File: pytorch_lightning/strategies/ddp2.py [hunk 3]
```diff
@@ -16,15 +16,31 @@
...
+            "The `DDP2Strategy` is deprecated in v1.7 and will be removed in v1.8. For single-node execution, we"
+            " recommend the `DDPStrategy` or the `DPStrategy`. If you need DDP2, you will need `torch < 1.9` and"
+            " `pytorch-lightning < 1.5`."
+        )
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/strategies/ddp2.py:42]: Yeah I wasn't sure. We could. Considering our stable API promise for 1.6, I thought it would be best to make it a deprecation only here in case someone subclasses and overrides a few methods. I don't see a strong need to remove the code for us, but I'm fine either way

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/strategies/ddp2.py:42]: > Considering our stable API promise for 1.6

Note that this is not something we really advertised in the release.

If we are moving forward with flattening, then doing this is interesting since it's one file less to update.

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/strategies/ddp2.py]: This is probably fine. Let's see what other reviewers think 👍

File: docs/source/accelerators/gpu_intermediate.rst [hunk 4]
```diff
@@ -104,6 +103,12 @@ In these situations you should use `dp` or `ddp_spawn` instead.
 
 Distributed Data Parallel 2
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. warning::
+    The DDP2 strategy is no longer supported. For single-node use, we recommend ``strategy='ddp'`` or
+    ``strategy='dp'`` as a replacement. If you need DDP2, you will need ``torch < 1.9``,
+    ``pytorch-lightning < 1.5``, and set it as ``accelerator='ddp2'``.
```
Comment (rohitgr7) [inline docs/source/accelerators/gpu_intermediate.rst:110]:
```suggestion
    ``pytorch-lightning < 1.6``, and set it as ``accelerator='ddp2'``.
```
it's present in v1.6 release?

File: docs/source/common/lightning_module.rst [hunk 5]
```diff
@@ -420,7 +420,7 @@ Validating with DataParallel
 ============================
 
 When training using a ``strategy`` that splits data from each batch across GPUs, sometimes you might
```
Comment (rohitgr7) [inline docs/source/common/lightning_module.rst:422]:
```suggestion
When validating using a ``strategy`` that splits data from each batch across GPUs, sometimes you might
```

File: pytorch_lightning/strategies/ddp2.py [hunk 6]
```diff
@@ -11,74 +11,23 @@
...
+    """DDP2 behaves like DP in one node, but synchronization across nodes behaves like in DDP.
 
-
-class DDP2Strategy(DDPStrategy):
-    """DDP2 behaves like DP in one node, but synchronization across nodes behaves like in DDP."""
+    .. deprecated:: v1.7
+        This strategy is no longer supported in v1.7 will be removed completely in v1.8. For single-node execution, we
+        recommend the :class:`~pytorch_lightning.strategies.ddp.DDPStrategy` or the
+        :class:`~pytorch_lightning.strategies.dp.DataParallelStrategy` as a replacement. If you rely on DDP2, you will
+        need ``torch < 1.9`` and ``pytorch-lightning < 1.5``.
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/ddp2.py:23]: same here

File: pytorch_lightning/strategies/ddp2.py [hunk 7]
```diff
@@ -11,74 +11,23 @@
...
-from pytorch_lightning.utilities.apply_func import apply_to_collection
-from pytorch_lightning.utilities.types import _METRIC_COLLECTION
+class DDP2Strategy:
+    """DDP2 behaves like DP in one node, but synchronization across nodes behaves like in DDP.
 
-
-class DDP2Strategy(DDPStrategy):
-    """DDP2 behaves like DP in one node, but synchronization across nodes behaves like in DDP."""
+    .. deprecated:: v1.7
+        This strategy is no longer supported in v1.7 will be removed completely in v1.8. For single-node execution, we
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/ddp2.py:20]: why one cycle deprecation?

File: pytorch_lightning/strategies/ddp2.py [hunk 8]
```diff
@@ -11,74 +11,23 @@
...
-    def register_strategies(cls, strategy_registry: Dict) -> None:
-        strategy_registry.register(
-            cls.strategy_name,
-            cls,
-            description=f"{cls.__class__.__name__}",
+    def __new__(cls, *args, **kwargs) -> None:
+        raise TypeError(
+            "The `DDP2Strategy`/`DDP2Plugin` is no longer supported in v1.7 and will be removed completely in v1.8."
+            " For single-node execution, we recommend the `DDPStrategy` or the `DPStrategy`. If you rely on DDP2, you"
+            " will need `torch < 1.9` and `pytorch-lightning < 1.5`."
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/ddp2.py:32]: same here

File: tests/accelerators/test_accelerator_connector.py [hunk 9]
```diff
@@ -340,7 +336,7 @@ def test_set_devices_if_none_cpu():
     assert trainer.num_devices == 3
 
 
-@pytest.mark.parametrize("strategy", ["ddp2", "dp"])
+@pytest.mark.parametrize("strategy", ["dp"])
```
Comment (rohitgr7) [inline tests/accelerators/test_accelerator_connector.py:339]: I think parametrization can be removed since there's only one left now.

File: pytorch_lightning/strategies/ddp2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/strategies/ddp2.py]: I'd also opt for removing that directly. Since it is broken anyways and we didn't get any complains, so nobody is using it :D

File: docs/source/accelerators/gpu_intermediate.rst (hunk 4 above)
Comment (awaelchli) [reply docs/source/accelerators/gpu_intermediate.rst:110]: It has to be PL < 1.5. The feature is broken since then.

Comment (rohitgr7) [reply docs/source/accelerators/gpu_intermediate.rst:110]: oh okay!

**Review by ananthsub — APPROVED** [review]

File: docs/source/accelerators/gpu_intermediate.rst [hunk 10]
```diff
@@ -418,7 +424,7 @@ To illustrate why this is needed, let's look at DataParallel
         x, y = batch
         y_hat = self(batch)
 
-        # on dp or ddp2 if we did softmax now it would be wrong
+        # on dp if we did softmax now it would be wrong
```
Comment (justusschock) [inline docs/source/accelerators/gpu_intermediate.rst:427]: I know, you just adapted it, but I don't get this, tbh. Shouldn't softmax be per sample of the batch? And since DP splits across batchdim, the samples should all be complete?
Maybe I am just being dumb right now...

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 11]
```diff
@@ -631,6 +631,13 @@ def _init_strategy(self) -> None:
             # TODO lazy initialized and setup horovod strategy `global_rank`
             self._handle_horovod()
         if isinstance(self._strategy_flag, str):
+            if self._strategy_flag == "ddp2":
+                # TODO: remove this error in v1.8
+                raise ValueError(
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:636]: Why do we need this error when it will fail on instantiation anyways?

**Review by carmocca — APPROVED** [review]

File: tests/accelerators/test_accelerator_connector.py (hunk 9 above)
Comment (awaelchli) [reply tests/accelerators/test_accelerator_connector.py:339]: updated

**Review by akihironitta — APPROVED** [review]

File: docs/source/accelerators/gpu_faq.rst [hunk 12]
```diff
@@ -36,25 +36,25 @@ In DDP, DDP_SPAWN, Deepspeed, DDP_SHARDED, or Horovod your effective batch size
...
 .. code-block:: python
 
     # effective batch size = 7
-    Trainer(accelerator="gpu", devices=8, strategy="ddp2")
     Trainer(accelerator="gpu", devices=8, strategy="dp")
 
     # effective batch size = 7 * 10
-    Trainer(accelerator="gpu", devices=8, num_nodes=10, strategy="ddp2")
-    Trainer(accelerator="gpu", devices=8, strategy="dp")
+    Trainer(accelerator="gpu", devices=8, num_nodes=10, strategy="dp")
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/accelerators/gpu_faq.rst]: Yes.

File: docs/source/accelerators/gpu_intermediate.rst (hunk 10 above)
Comment (awaelchli) [reply docs/source/accelerators/gpu_intermediate.rst:427]: @williamFalcon wrote this. I can imagine that for things like contrastive learning where both positive and negative samples are included in the batch, you would take multiple of them and combine them in the computational graph together. I do not know if that's what this code is supposed to show.

Comment (justusschock) [reply docs/source/accelerators/gpu_intermediate.rst:427]: Could be, but I guess, for 99% of the use-cases this is actually not relevant which is why I would remove this to avoid confusion here.

**Review by Borda — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Neat !
